roles/upgrade/post-upgrade/tasks/main.yml

```diff
@@ -11,6 +11,15 @@
     --timeout={{ upgrade_post_cilium_wait_timeout }}
   delegate_to: "{{ groups['kube_control_plane'][0] }}"
 
+- name: Ensure control-plane taints after upgrade
```

Comment:
This task is skipped for every `kube_node`, Notice the `not` in the second conditional. So this task won't taint every `kube_node`, it just taint all `kube_control_plane` except when a node is both `kube_control_plane` and `kube_node`. In this case the tant should not be set in order to allow the node to run all types of load, since it's both `worker` and `control-plane` in this case.

If a node is just a `kube_control_plane` the user should not have the choice to not taint it, since it's the default configuration for k8s `control-planes`. If the user wants to install a `control-plane` which can run any type of loads, without any toleration configuration, it should add it to both `kube_control_plane` and `kube_node` groups. This is already documented in https://github.com/kubernetes-sigs/kubespray/blob/master/docs/ansible.md#inventory.

> If you want the server to act both as control-plane and node, the server must be defined on both groups kube_control_plane and kube_node. If you want a standalone and unschedulable control plane, the server must be defined only in the kube_control_plane and not kube_node.

Finally I don't think there is any need of doing it somewhat early in order to allow toleration adaptations. As I have said in https://github.com/kubernetes-sigs/kubespray/pull/10464#issuecomment-1733296771, this taint migration won't break any deployment which defines both tolerations.

After the upgrade, the user will be able to update their deployments to remove the legacy not needed toleration, but this can be done after the upgrade.

Reply:
> This task is skipped for every kube_node, notice the not in the second conditional. 
Ah yes indeed my bad!

> If a node is just a kube_control_plane the user should not have the choice to not taint it, since it's the default configuration for k8s control-planes. If the user wants to install a control-plane which can run any type of loads, without any toleration configuration, it should add it to both kube_control_plane and kube_node groups. This is already documented in https://github.com/kubernetes-sigs/kubespray/blob/master/docs/ansible.md#inventory.

Yes but I am bit concerned by this because some users might have untaint a node outside of kubespray scope. And I am not sure if we should force the taints on users. basically.

> Finally I don't think there is any need of doing it somewhat early in order to allow toleration adaptations. As I have said in https://github.com/kubernetes-sigs/kubespray/pull/10464#issuecomment-1733296771, this taint migration won't break any deployment which defines both tolerations.
> 
> After the upgrade, the user will be able to update their deployments to remove the legacy not needed toleration, but this can be done after the upgrade.

I know but doing this early would allow us to remove our legacy taint safely within one kubespray version. If we don't do that now, we will have to wait for 1 maybe 2 kubespray release to do that and this might get lost in the process. So I think that having this taint early and removing the legacy taint everywhere else in kubespray would very much help

Reply:
> Yes but I am bit concerned by this because some users might have untaint a node outside of kubespray scope. And I am not sure if we should force the taints on users. basically.

The thing is that kubeadm removes the legacy taint `node-role.kubernetes.io/master` during the upgrade to `v1.25`. So in case of a cluster where `control-plane` nodes just have this taint and not both (legacy and current taints), we would need to ensure that `node-role.kubernetes.io/control-plane` is set before the upgrade. Do you think that adding the `node-role.kubernetes.io/control-plane` taint before the upgrade in case of `node-role.kubernetes.io/master` taint is set is a better approach?

I think that it could be cool in order to cover some corner cases (like some users who might have untaint a node outside of kubespray scope) altough this corner case should not be covered, as they are not following the Kubespray guidelines to configure a shared `control-plane` `worker` node  IMHO.

I just decided to set the taint just before uncordoning the node in order to ensure idempotency and code readability. But if you think that covering these corner cases are important enough I can check how to implement it as nice as it's now.

Reply:
> The thing is that kubeadm removes the legacy taint node-role.kubernetes.io/master during the upgrade to v1.25. So in case of a cluster where control-plane nodes just have this taint and not both (legacy and current taints), we would need to ensure that node-role.kubernetes.io/control-plane is set before the upgrade. Do you think that adding the node-role.kubernetes.io/control-plane taint before the upgrade in case of node-role.kubernetes.io/master taint is set is a better approach?

Kubernetes 1.25 is no longer supported in master so we shouldn't need to worry about the master role if it is already removed in that version then and we can get rid of it everywhere. Considering that we don't support 1.25 I am not sure where we should add the control-plane taint though :thinking:. I see two possible paths moving forward:

- Backport the code enforcing the control plane taint to kubespray 2.23.0 + add a note to kubespray 2.24.0 mentioning that users that did a 1.25 upgrade and didn't upgraded their cluster once using the new 2.23 version (2.23.1 probably?) should check their taint and that the control-plane should be re-added manually
- Systematically enforce control plane node but that's technically a change of behavior vs what we used to do. As before we didn't enforce that on upgrade, so people could remove the taint outside of kubespray and it would work even when reapplying kubespray afaiu

Reply:
Since the aim of this PR is fixing https://github.com/kubernetes-sigs/kubespray/issues/10217 I think that we can do the following:

- `master`: Just keep in the changes in the `kubeadm` configuration files to remove the legacy taint in fresh installations. Also we can include the changes suggested in https://github.com/kubernetes-sigs/kubespray/pull/10464#issuecomment-1758395518 since every cluster installed using the future 2.24.X won have the legacy taint and every cluster upgraded using the future 2.24.X will be already with k8s 1.25.

- `release-2.23`: Include the option to enforcing the control plane taint in order to avoid missing all control plane taints if upgrading to k8s 1.25. In this case we can tune this PR in orde to avoid this change of behavior about force tainting any node during the upgrade.  Also I think is a good idea mentioning the users taht have already upgraded their clusters to k8s v1.25 with any current Kubespray version as you have said.

What do you think about this?

Thanks for the feedback :smile:

Reply:
That would be perfect thank you🙏.

And thanks for bearing with me I admit that I kind off multitasked with other topics and didn't get every details in my previous answers. We do have a nice solution now though 🤝! :D

Reply:
> That would be perfect thank you🙏.
> 
> And thanks for bearing with me I admit that I kind of multitasked with other topics and didn't get every details in my previous answers. We do have a nice solution now though 🤝! :D

I'll try to make this changes next week. I'll ping you as soon as they are ready.

No problem, I appreciate the feedback. I also multitask more than I want to, I know the feeling :smile: